#Day 5 Part 1
inputs = d5_input.txt

--- Day 5: Print Queue ---
Satisfied with their search on Ceres, the squadron of scholars suggests subsequently scanning the stationery stacks of sub-basement 17.

The North Pole printing department is busier than ever this close to Christmas, and while The Historians continue their search of this historically significant facility, an Elf operating a very familiar printer beckons you over.

The Elf must recognize you, because they waste no time explaining that the new sleigh launch safety manual updates won't print correctly. Failure to update the safety manuals would be dire indeed, so you offer your services.

Safety protocols clearly indicate that new pages for the safety manuals must be printed in a very specific order. The notation X|Y means that if both page number X and page number Y are to be produced as part of an update, page number X must be printed at some point before page number Y.

The Elf has for you both the page ordering rules and the pages to produce in each update (your puzzle input), but can't figure out whether each update has the pages in the right order.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

file_path = '/content/drive/MyDrive/Advent 2024/day5/d5_input.txt'

import numpy as np
import pandas as pd

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


##example data

The first section specifies the **page ordering rules**, one per line. The first rule, 47|53, means that if an update includes both page number 47 and page number 53, then page number 47 must be printed at some point before page number 53. (47 doesn't necessarily need to be immediately before 53; other pages are allowed to be between them.)

In [ ]:
##example data
import pandas as pd

data_str = """47|53
97|13
97|61
97|47
75|29
61|13
75|53
29|13
97|29
53|29
61|53
97|53
61|29
47|13
75|47
97|75
47|61
75|61
47|29
75|13
53|13
"""

pages = """75,47,61,53,29
97,61,53,29,13
75,29,13
75,97,47,61,53
61,13,29
97,13,75,29,47
"""

The second section specifies the **page numbers of each update**. Because most safety manuals are different, the pages needed in the updates are different too. The first update, 75,47,61,53,29, means that the update consists of page numbers 75, 47, 61, 53, and 29.

To get the printers going as soon as possible, **start by identifying which updates are already in the right order.**

In the above example, the first update (75,47,61,53,29) is in the right order:

75 is correctly first because there are rules that put each other page after it: 75|47, 75|61, 75|53, and 75|29.
47 is correctly second because 75 must be before it (75|47) and every other page must be after it according to 47|61, 47|53, and 47|29.
61 is correctly in the middle because 75 and 47 are before it (75|61 and 47|61) and 53 and 29 are after it (61|53 and 61|29).
53 is correctly fourth because it is before page number 29 (53|29).
29 is the only page left and so is correctly last.

In [ ]:
rules = set()

for line in data_str.strip().split('\n'):
  a, b = map(int, line.split('|'))
  rules.add((a, b))


updates = []
for line in pages.strip().split('\n'):
  updates.append([int(x) for x in line.split(',')])

#example rules and updates are defined

In [ ]:
print(rules)

{(47, 53), (97, 75), (47, 13), (97, 29), (53, 29), (97, 47), (97, 53), (75, 29), (53, 13), (97, 13), (29, 13), (75, 47), (47, 61), (75, 53), (75, 13), (61, 29), (97, 61), (61, 53), (75, 61), (61, 13), (47, 29)}


In [ ]:
print(updates)

[[75, 47, 61, 53, 29], [97, 61, 53, 29, 13], [75, 29, 13], [75, 97, 47, 61, 53], [61, 13, 29], [97, 13, 75, 29, 47]]


**correct_order(updates, rules)**

adds value by value to a set

checks each value within the set for pairs (X, Y) that have known rules

if known rule is found compare correct (rule) index vs that in list

return False if fails index check

else return True if none fail


**get_middle_value(update)**

returns the middle value from a list

In [ ]:
def correct_order(updates, rules):
    seen = set()
    for num_val in updates: #add num_val to set
        seen.add(num_val)
        for a, b in rules:
            if a in seen and b in seen: #is (a,b) present?
                if updates.index(b) < updates.index(a): #aka is B before A? if so return false
                    return False
    return True

def get_middle_value(update):
    return update[len(update) // 2]

In [ ]:
def update_order(updates,rules):
    seen = set()
    all_updated_list_combos = []
    for num_val in updates: #add num_val to set
        seen.add(num_val)
        for a, b in rules:
            if a in seen and b in seen: #is (a,b) present?
                if updates.index(b) < updates.index(a): #aka is B before A? if so return false
                    #new copy each time we encounter a rule
                    updates[updates.index(a)], updates[updates.index(b)] = updates[updates.index(b)], updates[updates.index(a)]
                    return updates

part 1 example

In [ ]:
total = 0
import itertools
#part 1 only
for update in updates: #go line by line
    if correct_order(update, rules): #is line in correct order?
        total += get_middle_value(update) #grab middle index
print(total) #143

143


part 2 example

In [ ]:
total, incorrect_total = 0, 0
import itertools

for update in updates: #go line by line
    if correct_order(update, rules): #is line in correct order?
        total += get_middle_value(update) #grab middle index
    else:
      correct = False
      while correct == False:
        # print('incorrect update', update)
        updated_order = update_order(update,rules)
        # print('testing updated order', updated_order)

        if correct_order(updated_order, rules):
          # print('corrected update', updated_order)

          correct =True
          incorrect_total += get_middle_value(updated_order) #grab middle index

print(total, incorrect_total) #143

incorrect update [75, 97, 47, 61, 53]
testing updated order [97, 75, 47, 61, 53]
corrected update [97, 75, 47, 61, 53]
incorrect update [61, 13, 29]
testing updated order [61, 29, 13]
corrected update [61, 29, 13]
incorrect update [97, 13, 75, 29, 47]
testing updated order [97, 75, 13, 29, 47]
incorrect update [97, 75, 13, 29, 47]
testing updated order [97, 75, 29, 13, 47]
incorrect update [97, 75, 29, 13, 47]
testing updated order [97, 75, 29, 47, 13]
incorrect update [97, 75, 29, 47, 13]
testing updated order [97, 75, 47, 29, 13]
corrected update [97, 75, 47, 29, 13]
143 123


part 1 real

In [ ]:
updates, rules = [], set()

for line in open(file_path):
    line = line.strip()

    if ',' in line: #is update
        updates.append([int(x) for x in line.split(',')])

    if '|' in line: #is rule
        a, b = map(int, line.split('|'))
        rules.add((a, b))

total = 0
import itertools
#part 1 only
for update in updates: #go line by line
    if correct_order(update, rules): #is line in correct order?
        total += get_middle_value(update) #grab middle index
print(total) #4814


4814


part 2 example

--- Part Two ---
While the Elves get to work printing the correctly-ordered updates, you have a little time to fix the rest of them.

**For each of the incorrectly-ordered updates**, use the page ordering rules to put the page numbers in the right order. For the above example, here are the three incorrectly-ordered updates and their correct orderings:

75,97,47,61,53 becomes 97,75,47,61,53.
61,13,29 becomes 61,29,13.
97,13,75,29,47 becomes 97,75,47,29,13.
After taking only the incorrectly-ordered updates and ordering them correctly, their middle page numbers are 47, 29, and 47. Adding these together produces 123.

Find the updates which are not in the correct order. What do you get if you add up the middle page numbers after correctly ordering just those updates?

In [ ]:
updates, rules = [], set()

for line in open(file_path):
    line = line.strip()

    if ',' in line: #is update
        updates.append([int(x) for x in line.split(',')])

    if '|' in line: #is rule
        a, b = map(int, line.split('|'))
        rules.add((a, b))



total, incorrect_total = 0, 0
import itertools

for update in updates: #go line by line
    if correct_order(update, rules): #is line in correct order?
        total += get_middle_value(update) #grab middle index
    else:
      correct = False
      while correct == False:
        # print('incorrect update', update)
        updated_order = update_order(update,rules)
        # print('testing updated order', updated_order)

        if correct_order(updated_order, rules):
          # print('corrected update', updated_order)

          correct =True
          incorrect_total += get_middle_value(updated_order) #grab middle index

print(total, incorrect_total) #4814 5448

4814 5448
